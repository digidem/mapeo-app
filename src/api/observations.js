// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
