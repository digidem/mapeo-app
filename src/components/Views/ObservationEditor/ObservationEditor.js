// @flow
import React from 'react';
import {
  View,
  Animated,
  Keyboard,
  KeyboardAvoidingView,
  Text,
  StyleSheet,
  TouchableOpacity,
  TextInput,
  Image,
  ImageBackground,
  FlatList,
  ScrollView
} from 'react-native';
import { withNavigationFocus } from 'react-navigation';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';
import Icon from 'react-native-vector-icons/MaterialIcons';
import CheckIcon from 'react-native-vector-icons/Octicons';
import CloseIcon from 'react-native-vector-icons/MaterialCommunityIcons';
import FeatherIcon from 'react-native-vector-icons/Feather';
import LocationPin from 'react-native-vector-icons/Entypo';
import I18n from 'react-native-i18n';
import type { UpdateRequest } from '@api/observations';
import type { Category } from '../../../types/category';
import type { Observation } from '../../../types/observation';
import CategoryPin from '../../../images/category-pin.png';
import PencilIcon from '../../../images/editor-details.png';
import {
  LIGHT_GREY,
  WHITE,
  MANGO,
  MEDIUM_GREY,
  VERY_LIGHT_BLUE
} from '../../../lib/styles';
import Header from '../../Base/Header';

export type StateProps = {
  category?: Category,
  selectedObservation?: Observation,
  observations: Observation[]
};

export type Props = {
  isFocused: boolean,
  navigation: any
};

export type DispatchProps = {
  updateObservation: (o: UpdateRequest) => void,
  goToPhotoView: (photoSource: string) => void,
  goToObservationFields: () => void,
  goToCameraView: () => void,
  goToCategories: () => void,
  goBack: () => void,
  goToTabBarNavigation: () => void,
  showSavedModal: () => void
};

type State = {
  goToCamera: boolean,
  keyboardShown: boolean,
  text: string,
  keyboardHeight: number
};

const styles = StyleSheet.create({
  bottomButton: {
    alignSelf: 'stretch',
    justifyContent: 'center',
    backgroundColor: WHITE,
    height: 60,
    borderColor: LIGHT_GREY,
    borderBottomWidth: 1
  },
  bottomButtonText: {
    fontSize: 20,
    fontWeight: '700',
    color: 'black'
  },
  categoryAtText: {
    fontSize: 12,
    color: 'black',
    fontWeight: '400'
  },
  categoryName: {
    fontSize: 15,
    color: 'black',
    fontWeight: '600'
  },
  categoryPin: {
    width: 80,
    height: 90,
    marginTop: 5,
    justifyContent: 'center',
    alignItems: 'center'
  },
  categoryPositionText: {
    fontSize: 12,
    color: 'black',
    fontWeight: '700',
    marginLeft: 3
  },
  check: {
    backgroundColor: MANGO,
    height: 25,
    width: 25,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  },
  checkOuterCircle: {
    width: 30,
    height: 30,
    backgroundColor: '#ed6109',
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  },
  checkIcon: {
    alignSelf: 'center',
    marginLeft: 3
  },
  circle: {
    width: 60,
    height: 60,
    backgroundColor: 'white',
    borderRadius: 50,
    borderColor: LIGHT_GREY,
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: 'black',
    shadowRadius: 10,
    shadowOffset: { width: 0, height: 2 },
    elevation: 3,
    shadowOpacity: 1,
    margin: 5
  },
  collectionsImg: {
    alignSelf: 'center'
  },
  container: {
    flex: 1,
    backgroundColor: WHITE
  },
  greyCheck: {
    backgroundColor: LIGHT_GREY,
    height: 25,
    width: 25,
    borderRadius: 50,
    justifyContent: 'center'
  },
  greyCheckOuterCircle: {
    width: 30,
    height: 30,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#d6d2cf'
  },
  mediaRow: {
    backgroundColor: WHITE,
    borderColor: LIGHT_GREY,
    borderTopWidth: 1,
    height: 85
  },
  mediaRowKeyboardShown: {
    flex: 1,
    backgroundColor: WHITE,
    borderColor: LIGHT_GREY,
    marginBottom: -100
  },
  photosButton: {
    alignSelf: 'stretch',
    justifyContent: 'center',
    backgroundColor: WHITE,
    height: 60,
    borderColor: LIGHT_GREY,
    borderBottomWidth: 1
  },
  textInput: {
    flex: 1,
    fontSize: 20,
    padding: 20,
    paddingBottom: 30,
    color: 'black',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
    textAlignVertical: 'top',
    backgroundColor: 'white',
    borderColor: LIGHT_GREY
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    color: 'black',
    textAlign: 'center'
  },
  titleLong: {
    fontSize: 18,
    fontWeight: '700',
    color: 'black',
    textAlign: 'center'
  },
  triangle: {
    alignSelf: 'center',
    width: 0,
    height: 0,
    backgroundColor: 'transparent',
    borderStyle: 'solid',
    borderLeftWidth: 7,
    borderRightWidth: 7,
    borderBottomWidth: 7,
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'rgba(0, 0, 0, .8)',
    transform: [{ rotate: '180deg' }]
  }
});

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

class ObservationEditor extends React.Component<
  Props & StateProps & DispatchProps,
  State
> {
  paddingInput: Animated.Value;
  keyboardWillShowListener: any;
  keyboardWillHideListener: any;
  keyboardDidShowListener: any;
  keyboardDidHideListener: any;
  scrollView: any;
  textInput: any;

  constructor(props: StateProps & DispatchProps) {
    super();

    this.paddingInput = new Animated.Value(0);

    this.state = {
      keyboardShown: false,
      text: props.selectedObservation ? props.selectedObservation.notes : '',
      goToCamera: false,
      keyboardHeight: 0
    };
  }

  componentWillMount() {
    this.keyboardWillShowListener = Keyboard.addListener(
      'keyboardWillShow',
      this.keyboardWillShow
    );
    this.keyboardWillHideListener = Keyboard.addListener(
      'keyboardWillHide',
      this.keyboardWillHide
    );
  }

  componentDidMount() {
    const { updateObservation, selectedObservation, category } = this.props;

    if (selectedObservation && category) {
      updateObservation({
        ...selectedObservation,
        type: category.name,
        name: category.name
      });
    }

    this.keyboardDidShowListener = Keyboard.addListener(
      'keyboardDidShow',
      this.keyboardDidShow
    );
    this.keyboardDidHideListener = Keyboard.addListener(
      'keyboardDidHide',
      this.keyboardDidHide
    );
  }

  shouldComponentUpdate(
    nextProps: Props & StateProps & DispatchProps,
    nextState: State
  ) {
    if (nextProps.isFocused) {
      return nextProps !== this.props || nextState !== this.state;
    }

    return false;
  }

  componentWillUnmount() {
    this.keyboardWillShowListener.remove();
    this.keyboardWillHideListener.remove();
    this.keyboardDidShowListener.remove();
    this.keyboardDidHideListener.remove();
  }

  keyboardWillShow = e => {
    Animated.timing(this.paddingInput, {
      duration: e.duration,
      toValue: 60
    }).start();
  };

  keyboardWillHide = e => {
    Animated.timing(this.paddingInput, {
      duration: e.duration,
      toValue: 0
    }).start();
  };

  keyboardDidShow = e => {
    this.setState(previousState => ({
      goToCamera: false,
      keyboardShown: true,
      text: previousState.text,
      keyboardHeight: e.endCoordinates.height
    }));
  };

  keyboardDidHide = () => {
    if (this.state.goToCamera) {
      this.setState(previousState => ({
        goToCamera: false,
        keyboardShown: false,
        text: previousState.text
      }));
      this.props.goToCameraView();
    } else {
      this.setState(previousState => ({
        goToCamera: previousState.goToCamera,
        keyboardShown: false,
        text: previousState.text
      }));
    }
  };

  handleTextInputChange = text => {
    this.setState({ text });
  };

  handleSaveObservation = () => {
    const {
      updateObservation,
      selectedObservation,
      goToTabBarNavigation,
      showSavedModal
    } = this.props;
    const { text } = this.state;

    if (selectedObservation) {
      updateObservation({
        ...selectedObservation,
        notes: text
      });
    }

    showSavedModal();
    goToTabBarNavigation();
  };

  handleTextInputBlur = () => {
    const { selectedObservation, updateObservation } = this.props;

    if (selectedObservation) {
      updateObservation({
        id: selectedObservation.id,
        notes: this.state.text
      });
    }
  };

  goToCameraView = () => {
    const {
      selectedObservation,
      updateObservation,
      goToCameraView
    } = this.props;

    if (selectedObservation) {
      updateObservation({
        id: selectedObservation.id,
        notes: this.state.text
      });
    }
    if (this.state.keyboardShown) {
      this.setState(previousState => ({
        goToCamera: true,
        keyboardShown: false,
        text: previousState.text
      }));
    } else {
      goToCameraView();
    }
    Keyboard.dismiss();
  };

  goToCategoriesView = () => {
    const {
      goBack,
      goToCategories,
      observations,
      selectedObservation
    } = this.props;
    let updateFlow = false;
    if (selectedObservation) {
      observations.forEach(o => {
        if (o.id === selectedObservation.id) {
          updateFlow = true;
        }
      });
    }

    if (updateFlow) {
      goToCategories();
    } else {
      goBack();
    }
  };

  goToObservationFields = () => {
    const { goToObservationFields } = this.props;
    goToObservationFields();
    Keyboard.dismiss();
  };

  render() {
    const {
      navigation,
      goBack,
      selectedObservation,
      goToPhotoView,
      goToObservationFields
    } = this.props;
    const { keyboardShown, text } = this.state;
    const positionText = selectedObservation
      ? `${selectedObservation.lat}, ${selectedObservation.lon}`
      : 'Loading...';
    const keyExtractor = item => item.source;

    let fieldAnswered;
    if (selectedObservation) {
      fieldAnswered = selectedObservation.fields.find(f => f.answered);
    }

    const showGreyCheck =
      text === '' &&
      selectedObservation &&
      !selectedObservation.media.length &&
      !fieldAnswered;

    if (!selectedObservation) {
      goBack();
      return <View />;
    }

    return (
      <KeyboardAvoidingView style={styles.container}>
        <Header
          leftIcon={
            <TouchableOpacity
              underlayColor="rgba(0, 0, 0, 0.5)"
              onPress={goBack}
            >
              <CloseIcon color="#9E9C9C" name="window-close" size={30} />
            </TouchableOpacity>
          }
          rightIcon={
            showGreyCheck ? (
              <View style={styles.greyCheckOuterCircle}>
                <View style={styles.greyCheck}>
                  <CheckIcon
                    color="white"
                    name="check"
                    size={18}
                    style={styles.checkIcon}
                  />
                </View>
              </View>
            ) : (
              <TouchableOpacity
                style={styles.checkOuterCircle}
                onPress={this.handleSaveObservation}
              >
                <View style={styles.check}>
                  <CheckIcon
                    color="white"
                    name="check"
                    size={18}
                    style={styles.checkIcon}
                  />
                </View>
              </TouchableOpacity>
            )
          }
          showTriangle
          style={{
            backgroundColor: VERY_LIGHT_BLUE
          }}
        />
        <View
          style={{
            justifyContent: 'center',
            alignItems: 'center',
            paddingBottom: 10,
            backgroundColor: VERY_LIGHT_BLUE
          }}
        >
          <TouchableOpacity onPress={this.goToCategoriesView}>
            <View style={styles.circle}>
              {selectedObservation.icon && (
                <Image
                  source={selectedObservation.icon}
                  style={{ width: 30, height: 30 }}
                  resizeMode="contain"
                />
              )}
            </View>
          </TouchableOpacity>
          <View style={{ marginTop: 5, flexDirection: 'row' }}>
            <LocationPin color={MANGO} name="location-pin" size={15} />
            <Text style={styles.categoryPositionText}>{positionText}</Text>
          </View>
        </View>
        <ScrollView
          ref={ref => (this.scrollView = ref)}
          contentContainerStyle={{ flex: 1 }}
        >
          <TextInput
            ref={ref => (this.textInput = ref)}
            style={styles.textInput}
            value={text}
            onChangeText={this.handleTextInputChange}
            onFocus={() => this.scrollView.scrollToEnd()}
            placeholder={I18n.t('editor.placeholder')}
            placeholderTextColor="silver"
            underlineColorAndroid="transparent"
            onBlur={this.handleTextInputBlur}
            multiline
            autoGrow
            autoFocus
          />
        </ScrollView>
        {!!selectedObservation &&
          !!selectedObservation.media.length && (
            <View
              style={
                keyboardShown ? styles.mediaRowKeyboardShown : styles.mediaRow
              }
            >
              <FlatList
                horizontal
                scrollEnabled
                style={{
                  flex: 1,
                  flexDirection: 'row',
                  paddingTop: 10
                }}
                contentContainerStyle={{
                  alignContent: 'flex-start'
                }}
                keyExtractor={keyExtractor}
                renderItem={({ item }) => (
                  <TouchableOpacity
                    onPress={() => goToPhotoView(item.source)}
                    style={{ paddingLeft: 10 }}
                  >
                    <Image
                      source={{ uri: item.source }}
                      style={{
                        width: 65,
                        height: 65,
                        borderRadius: 5
                      }}
                    />
                  </TouchableOpacity>
                )}
                data={selectedObservation.media}
              />
            </View>
          )}
        <View
          style={{
            borderTopWidth: 1,
            borderColor: LIGHT_GREY
          }}
        >
          {keyboardShown && (
            <Animated.View
              style={{
                flexDirection: 'row',
                alignItems: 'center',
                justifyContent: 'space-around',
                backgroundColor: 'white',
                marginBottom: this.state.keyboardHeight,
                height: 55
              }}
            >
              <TouchableOpacity
                onPress={this.goToCameraView}
                underlayColor="transparent"
                style={{ flex: 1, alignItems: 'center' }}
              >
                <Icon color={MEDIUM_GREY} name="photo-camera" size={30} />
              </TouchableOpacity>
              <TouchableOpacity
                onPress={this.goToObservationFields}
                underlayColor="transparent"
                style={{ flex: 1, alignItems: 'center' }}
              >
                <Image source={PencilIcon} style={{ width: 25, height: 25 }} />
              </TouchableOpacity>
            </Animated.View>
          )}
          {!keyboardShown && (
            <View>
              <TouchableOpacity
                style={styles.photosButton}
                onPress={this.goToCameraView}
                underlayColor="transparent"
              >
                <View style={{ flexDirection: 'row' }}>
                  <Icon
                    color={MEDIUM_GREY}
                    name="photo-camera"
                    size={30}
                    style={{ marginHorizontal: 30 }}
                  />
                  <Text style={styles.bottomButtonText}>
                    {I18n.t('editor.media_button')}
                  </Text>
                </View>
              </TouchableOpacity>
              <TouchableOpacity
                style={styles.bottomButton}
                onPress={this.goToObservationFields}
                underlayColor="transparent"
              >
                <View style={{ flexDirection: 'row' }}>
                  <Image
                    source={PencilIcon}
                    style={{
                      marginLeft: 34,
                      marginRight: 31,
                      width: 25,
                      height: 25
                    }}
                  />
                  <Text style={styles.bottomButtonText}>
                    {I18n.t('editor.details_button')}
                  </Text>
                </View>
              </TouchableOpacity>
            </View>
          )}
        </View>
      </KeyboardAvoidingView>
    );
  }
}

export default withNavigationFocus(ObservationEditor);
