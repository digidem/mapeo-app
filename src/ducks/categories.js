// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: CATEGORY_LIST,
  action: categoryList,
  reducer: categoryListReducer
} = create('CATEGORY_LIST', {
  success: (state, meta, payload) => {
    const newState = update(state, {
      categories: { $set: keyBy(payload, 'id') }
    });

    return newState;
  }
});

export default [categoryListReducer];
