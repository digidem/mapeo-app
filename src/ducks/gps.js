// @flow

import update from 'immutability-helper';
import { create } from '../lib/redux';
import {
  resourcePending,
  resourceSuccess,
  resourceFailed
} from '../lib/resource';

export const {
  type: GPS_PENDING,
  action: gpsPending,
  reducer: gpsPendingReducer
} = create('GPS_PENDING', {
  start: (state, meta) => {
    const newState = update(state, {
      gps: { $set: resourcePending(meta) }
    });

    return newState;
  }
});

export const { type: GPS_SET, action: gpsSet, reducer: gpsSetReducer } = create(
  'GPS_SET',
  {
    start: (state, meta) => {
      const newState = update(state, {
        gps: { $set: resourceSuccess(meta) }
      });

      return newState;
    }
  }
);

export const {
  type: GPS_FAILED,
  action: gpsFailed,
  reducer: gpsFailedReducer
} = create('GPS_FAILED', {
  start: (state, meta) => {
    const newState = update(state, {
      gps: { $set: resourceFailed(meta) }
    });

    return newState;
  }
});

export default [gpsPendingReducer, gpsSetReducer, gpsFailedReducer];
