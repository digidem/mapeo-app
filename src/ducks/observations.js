// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
