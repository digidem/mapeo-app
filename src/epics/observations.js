// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
