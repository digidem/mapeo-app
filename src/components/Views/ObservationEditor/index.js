// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { values } from 'lodash';
import { withNavigationFocus } from 'react-navigation';
import { StoreState } from '../../../types/redux';

import {
  observationUpdate,
  observationSelect,
  observationSave,
  observationUpdateSave
} from '../../../ducks/observations';
import { modalHide, modalShow } from '../../../ducks/modals';
import ObservationEditor from './ObservationEditor';
import type { Props, StateProps, DispatchProps } from './ObservationEditor';

function mapStateToProps(state: StoreState, ownProps: Props): StateProps {
  const category = state.app.selectedObservation
    ? state.app.categories[state.app.selectedObservation.categoryId]
    : undefined;

  return {
    category:
      (state.app.categories &&
        ownProps.navigation.state &&
        ownProps.navigation.state.params &&
        ownProps.navigation.state.params.category &&
        state.app.categories[ownProps.navigation.state.params.category]) ||
      category,
    selectedObservation: state.app.selectedObservation,
    observations: values(state.app.observations),
    observationSource: state.app.observationSource.source,
    cancelModalVisible: state.app.modals.cancelled,
    gps: state.app.gps,
    manualGPSModalVisible: state.app.modals.manualGPS,
    gpsFormat: state.app.settings.gpsFormat,
    icons: state.app.icons
  };
}

function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
  return {
    updateObservation: observation => dispatch(observationUpdate(observation)),
    clearSelectedObservation: () => dispatch(observationSelect(undefined)),
    showCancelModal: () => dispatch(modalShow('cancelled')),
    hideCancelModal: () => dispatch(modalHide('cancelled')),
    hideManualGPSModal: () => dispatch(modalHide('manualGPS')),
    showManualGPSModal: () => dispatch(modalShow('manualGPS')),
    saveObservation: (update: boolean) => {
      if (update) {
        dispatch(observationUpdateSave(null));
      } else {
        dispatch(observationSave(null));
      }
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(
  withNavigationFocus(ObservationEditor)
);
