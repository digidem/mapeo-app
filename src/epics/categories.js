// @flow
import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import { Image } from 'react-native';
import { CATEGORY_LIST, categoryList } from '../ducks/categories';
import { PRESETS_SELECT, presetsSelect } from '../ducks/fields';
import { createCategory } from '../mocks/categories';
import type { Action } from '../types/redux';
import type { Category } from '../types/category';
import Presets from '../api/presets';

export const categoryListEpic = (
  action$: ActionsObservable<Action<string, Category[]>>
) =>
  action$
    .ofType(CATEGORY_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() => Presets.list().map(list => presetsSelect(list[0])));

export default [categoryListEpic];
