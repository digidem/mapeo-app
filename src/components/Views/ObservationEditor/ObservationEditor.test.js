// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import ObservationEditor from './ObservationEditor';
import { createObservation } from '../../../mocks/observations';
import { createNavigationScreenProp } from '../../../mocks/navigation';
import { resourceSuccess } from '../../../lib/resource';
import { createCategory } from '../../../mocks/categories';
import { createGPSState } from '../../../mocks/gps';

jest.mock('../../Base/CancelModal/CancelModal', () => () => null);
jest.mock('../../Base/ManualGPSModal', () => () => null);
jest.mock('../../../ducks/gps', () => ({
  HIGH_ACCURACY: 'HIGH_ACCURACY'
}));

describe('ObservationEditor tests', () => {
  const observation = createObservation();

  test('snapshot', () => {
    const tree = renderer
      .create(
        <ObservationEditor
          navigation={createNavigationScreenProp()}
          observations={[]}
          observationSource=""
          cancelModalVisible={false}
          gps={resourceSuccess(createGPSState())}
          manualGPSModalVisible={false}
          gpsFormat=""
          icons={{}}
          updateObservation={jest.fn()}
          showCancelModal={jest.fn()}
          hideCancelModal={jest.fn()}
          clearSelectedObservation={jest.fn()}
          hideManualGPSModal={jest.fn()}
          showManualGPSModal={jest.fn()}
          saveObservation={jest.fn()}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
