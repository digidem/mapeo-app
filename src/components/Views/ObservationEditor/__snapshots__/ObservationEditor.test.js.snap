// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservationEditor tests snapshot 1`] = `
<View
  enabled={true}
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ccffff",
        "justifyContent": "center",
        "paddingBottom": 10,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "#EAEAEA",
            "borderRadius": 50,
            "borderWidth": 1,
            "elevation": 3,
            "height": 60,
            "justifyContent": "center",
            "margin": 5,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 10,
            "width": 60,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#ff9933",
              "fontSize": 15,
            },
            undefined,
            Object {
              "fontFamily": "Entypo",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 12,
            "fontWeight": "700",
            "marginLeft": 3,
          }
        }
      >
        50.5, 50.5
      </Text>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <TextInput
        allowFontScaling={true}
        autoFocus={false}
        autoGrow={true}
        multiline={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="What's happening here?"
        placeholderTextColor="silver"
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "white",
            "borderColor": "#EAEAEA",
            "color": "black",
            "flex": 1,
            "fontSize": 20,
            "justifyContent": "flex-start",
            "padding": 20,
            "paddingBottom": 30,
            "textAlignVertical": "top",
          }
        }
        underlineColorAndroid="transparent"
        value="This is pretty bad. It smells pretty noxious, and the entire pond is affected."
      />
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "borderColor": "#EAEAEA",
        "borderTopWidth": 1,
      }
    }
  >
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#ffffff",
            "borderBottomWidth": 1,
            "borderColor": "#EAEAEA",
            "height": 60,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#666666",
                  "fontSize": 30,
                },
                Object {
                  "marginHorizontal": 30,
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "black",
                "fontSize": 20,
                "fontWeight": "700",
              }
            }
          >
            Photos & Videos
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#ffffff",
            "borderBottomWidth": 1,
            "borderColor": "#EAEAEA",
            "height": 60,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/editor-details.png",
              }
            }
            style={
              Object {
                "height": 25,
                "marginLeft": 34,
                "marginRight": 31,
                "width": 25,
              }
            }
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "black",
                "fontSize": 20,
                "fontWeight": "700",
              }
            }
          >
            Details
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
